File: Python/Lesson_213_permutation_in_string.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 213 -- Permutation in String
# Category   : Sliding Window
# Difficulty : Medium
# Study Plan : Day 107
# =============================================================
#
# QUESTION:
#   Return true if s2 contains a permutation of s1.
# =============================================================
def checkInclusion(s1,s2):
    if len(s1)>len(s2): return False
    a=[0]*26; b=[0]*26
    for c in s1: a[ord(c)-97]+=1
    for i,c in enumerate(s2):
        b[ord(c)-97]+=1
        if i>=len(s1): b[ord(s2[i-len(s1)])-97]-=1
        if a==b: return True
    return False
if __name__=="__main__":
    print(checkInclusion("ab","eidbaooo"))
    print(checkInclusion("ab","eidboaoo"))

File: CPP/Lesson_213_permutation_in_string.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 213 -- Permutation in String
// Category   : Sliding Window
// Difficulty : Medium
// Study Plan : Day 107
// =============================================================
//
// QUESTION:
//   Return true if s2 contains a permutation of s1.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
bool checkInclusion(string s1,string s2){if(s1.size()>s2.size())return false;vector<int> a(26),b(26);for(char c:s1)a[c-'a']++;for(int i=0;i<(int)s2.size();i++){b[s2[i]-'a']++;if(i>=(int)s1.size())b[s2[i-s1.size()]-'a']--;if(a==b)return true;}return false;}
int main(){cout<<checkInclusion("ab","eidbaooo")<<"\n"<<checkInclusion("ab","eidboaoo")<<"\n";}
